From 523aa08f51a88c59ad4b1f600f8ce3d122e9e289 Mon Sep 17 00:00:00 2001
From: Guilherme Janczak <guilherme.janczak@yandex.com>
Date: Sun, 25 Jul 2021 01:55:25 +0000
Subject: [PATCH 1/4] remove always true condition in if statement

---
 stest.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/stest.c b/stest.c
index 7a7b0bc..e27d3a5 100644
--- a/stest.c
+++ b/stest.c
@@ -84,7 +84,7 @@ main(int argc, char *argv[])
 	if (!argc) {
 		/* read list from stdin */
 		while ((n = getline(&line, &linesiz, stdin)) > 0) {
-			if (n && line[n - 1] == '\n')
+			if (line[n - 1] == '\n')
 				line[n - 1] = '\0';
 			test(line, line);
 		}
-- 
2.33.0

